docker/deploy/Dockerfile

```diff
@@ -0,0 +1,73 @@
+ARG BUILD_ENV
+
+FROM $BUILD_ENV as builder
+
+ENV MKL_LIB=/opt/intel/compilers_and_libraries/linux/mkl/lib/intel64
+
+RUN ldconfig
+
+# Build DLA-Future
+COPY . /DLA-Future
+
+RUN touch /usr/bin/srun && chmod +x /usr/bin/srun
+
+SHELL ["/bin/bash", "-c"]
+
+RUN mkdir DLA-Future-build && cd DLA-Future-build && \
+    CC=/usr/local/mpich/bin/mpicc CXX=/usr/local/mpich/bin/mpicxx cmake /DLA-Future \
+      -DMKL_ROOT=/opt/intel/compilers_and_libraries/linux/mkl \
+      -DCMAKE_BUILD_TYPE=Release \
+      -DDLAF_WITH_CUDA=ON \
+      -DDLAF_WITH_MKL=ON \
+      -DDLAF_WITH_TEST=ON \
+      -DDLAF_BUILD_MINIAPPS=OFF \
+      -DDLAF_MPI_PRESET=slurm \
+      -DMPIEXEC_NUMCORES=32 \
+      -DMPIEXEC_EXECUTABLE=srun \
+      -DMPIEXEC_PREFLAGS="--jobid=\$JOBID;unbuffer;sarus;run;-t;--mpi;\$IMAGE" \
+      -DCMAKE_INSTALL_PREFIX=/usr && \
+      source /DLA-Future/ci/sarusify.sh && \
+      echo "$SARUS_TEST_COMMANDS" > /root/run.sh && \
+      make -j$(nproc) && \
+      # Let's just bundle the libs to make the docker image small
+      # We have to copy a couple of MKL libs by hand
+      /root/libtree/libtree --strip --chrpath \
+        -d /root/DLA-Future.bundle \
+        $TEST_BINARIES \
+        ${MKL_LIB}/libmkl_tbb_thread.so \
+        ${MKL_LIB}/libmkl_intel_thread.so \
+        ${MKL_LIB}/libmkl_sequential.so \
+        ${MKL_LIB}/libmkl_def.so \
+        ${MKL_LIB}/libmkl_vml_def.so \
+        ${MKL_LIB}/libmkl_vml_cmpt.so \
+        ${MKL_LIB}/libmkl_mc.so \
+        ${MKL_LIB}/libmkl_vml_mc.so \
+        ${MKL_LIB}/libmkl_mc3.so \
+        ${MKL_LIB}/libmkl_vml_mc3.so \
+        ${MKL_LIB}/libmkl_avx.so \
+        ${MKL_LIB}/libmkl_vml_avx.so \
+        ${MKL_LIB}/libmkl_avx2.so \
+        ${MKL_LIB}/libmkl_vml_avx2.so \
+        ${MKL_LIB}/libmkl_avx512_mic.so \
+        ${MKL_LIB}/libmkl_vml_avx512_mic.so \
+        ${MKL_LIB}/libmkl_avx512.so \
+        ${MKL_LIB}/libmkl_vml_avx512.so \
+        ${MKL_LIB}/libmkl_core.so && \
```

Comment:
Can you please explain what this step does and why only a part of the mkl's libraries is present?

Reply:
The build environment docker image is huge (5.94GB uncompressed atm), so what I'm doing to reduce it in size is to copy over only essential binaries for the tests using https://github.com/haampie/libtree. It figures out like `ldd` what the dependent shared libs are, and copies them to a single folder. MKL however uses `dlopen` to load microarchitecture specific libraries, so you have to copy those implicit dependencies yourself by listing them like that. The resulting image is "only" 711MB uncompressed (mostly due to MKL...).

Currently it is somewhat important to keep images small, because we cannot build them on Daint yet and therefore build and store them temporarily elsewhere. That means there is overhead of creating tarballs, but also storage can be an issue at some point (currently Gitlab is used to store images, which has a 10GB limit per repo for their Docker registry)